numpy/ma/core.py

```diff
@@ -3493,6 +3493,10 @@ def _get_flat(self):
     def _set_flat(self, value):
         "Set a flattened version of self to value."
         y = self.ravel()
+        # TODO: This is messed up, since there is no guarantee for
+        # ravel to return views into either array or mask.
+        # If self._sharedmask, we still do want self._mask to change here:
+        y._sharedmask = False
```

Comment:
OK, removed this fix for now, to be "minimal". Als omaybe the warning gives at least some hint of the weird behaviour that goes on and it will be automatically fixed with the other stuff. (only the sharedmask bug, not the one about non-contiguous arrays).